### ISSUE TITLE
Noob Help please.

### ISSUE THREAD

**Issue by Daxcor69** [travisghansen/kubernetes-client-php#3]
I am doing a simple test to see if the client will talk to my cluster. 
`<?php
// need to auto load the dependencies
require_once __DIR__ . '/vendor/autoload.php';

declare(ticks = 1);
pcntl_signal(SIGINT, function () {
    exit(0);
});

$config = KubernetesClient\Config::BuildConfigFromFile();
$client = new KubernetesClient\Client($config);

$params = [
    'limit' => 1
];
$list = $client->createList('/api/v1/nodes', $params);

// get all
$items = $list->get();
var_dump($items);
?>`

I have /home/brad/.kube/config and can use kubectl just fine as user brad.  This is the same user I run my test: $ php7.4 index.php

The output of the test is:

`array(8) {
  ["kind"]=>  string(6) "Status"
  ["apiVersion"]=>  string(2) "v1"
  ["metadata"]=>   array(0) {
  }
  ["status"]=>  string(7) "Failure"
  ["message"]=>  string(109) "nodes is forbidden: User "system:anonymous" cannot list resource "nodes" in API group "" at the cluster scope"
  ["reason"]=>  string(9) "Forbidden"
  ["details"]=>
    array(1) {
    ["kind"]=>
    string(5) "nodes"
  }
  ["code"]=>
  int(403)
}`

Not sure why I don't have permission, when I can do it from the cli using   kubectl get nodes.

**Comment by travisghansen** [comment]
Seems odd. What do you get if you `var_dump($config->getParsedConfigFile());`? I'm wondering why it's not picking up the data appropriately?

Can you send the config file (obviously scrub sensitive data)? I want to see the structure of it as different k8s distros can do different things and the lib may not support some specific syntax you've got going on ATM.

**Comment by Daxcor69** [comment]
I am using Digital Ocean managed kubernetes.

`array(7) {
  ["apiVersion"]=>
  string(2) "v1"
  ["clusters"]=>
  array(1) {
    [0]=>
    array(2) {
      ["cluster"]=>
      array(2) {
        ["certificate-authority-data"]=>
        string(1540) "<redacted>"
        ["server"]=>
        string(67) "https://<redacted>.k8s.ondigitalocean.com"
      }
      ["name"]=>
      string(19) "do-sfo2-dev-cluster"
    }
  }
  ["contexts"]=>
  array(1) {
    [0]=>
    array(2) {
      ["context"]=>
      array(2) {
        ["cluster"]=>
        string(19) "do-sfo2-dev-cluster"
        ["user"]=>
        string(25) "do-sfo2-dev-cluster-admin"
      }
      ["name"]=>
      string(19) "do-sfo2-dev-cluster"
    }
  }
  ["current-context"]=>
  string(19) "do-sfo2-dev-cluster"
  ["kind"]=>
  string(6) "Config"
  ["preferences"]=>
  array(0) {
  }
  ["users"]=>
  array(1) {
    [0]=>
    array(2) {
      ["name"]=>
      string(25) "do-sfo2-dev-cluster-admin"
      ["user"]=>
      array(1) {
        ["exec"]=>
        array(4) {
          ["apiVersion"]=>
          string(36) "client.authentication.k8s.io/v1beta1"
          ["args"]=>
          array(7) {
            [0]=>
            string(10) "kubernetes"
            [1]=>
            string(7) "cluster"
            [2]=>
            string(10) "kubeconfig"
            [3]=>
            string(15) "exec-credential"
            [4]=>
            string(17) "--version=v1beta1"
            [5]=>
            string(17) "--context=default"
            [6]=>
            string(36) "<redacted>"
          }
          ["command"]=>
          string(5) "doctl"
          ["env"]=>
          NULL
        }
      }
    }
  }
}`

**Comment by travisghansen** [comment]
OK, that could be it...lemme digest this a bit and get back..

In the meantime, can you send the yaml version by chance so it's a bit more readable?

**Comment by Daxcor69** [comment]
apiVersion: v1

clusters:

- cluster:

    certificate-authority-data: redacted

    server: https://redacted.k8s.ondigitalocean.com

  name: do-sfo2-dev-cluster

contexts:

- context:

    cluster: do-sfo2-dev-cluster

    user: do-sfo2-dev-cluster-admin

  name: do-sfo2-dev-cluster

current-context: do-sfo2-dev-cluster

kind: Config

preferences: {}

users:

- name: do-sfo2-dev-cluster-admin

  user:

    exec:

      apiVersion: client.authentication.k8s.io/v1beta1

      args:

      - kubernetes

      - cluster

      - kubeconfig

      - exec-credential

      - --version=v1beta1

      - --context=default

      - redacted

      command: doctl

      env: null



*From:* Travis Glenn Hansen [mailto:notifications@github.com]
*Sent:* Tuesday, September 29, 2020 4:55 PM
*To:* travisghansen/kubernetes-client-php
*Cc:* Daxcor69; Author
*Subject:* Re: [travisghansen/kubernetes-client-php] Noob Help please. (#3)



OK, that could be it...lemme digest this a bit and get back..

In the meantime, can you send the yaml version by chance so it's a bit more
readable?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/travisghansen/kubernetes-client-php/issues/3#issuecomment-701054295>,
or unsubscribe

**Comment by travisghansen** [comment]
Yeah, I see the issue. It looks like there are 2 different variants at play and I haven't added support for how DO does it:

- https://banzaicloud.com/blog/kubeconfig-security/#auth-provider-plugins
- https://banzaicloud.com/blog/kubeconfig-security/#exec-helper

Could you run the following and send over the scrubbed output?

```
doctl kubernetes cluster kubeconfig exec-credential --version=v1beta1 --context=default <redacted>
```

I'm not entirely sure what was in `<redacted>`.

In short what happens is a local command is executed by kubectl to get a token/info before submitting the request to DO. Typically that data will be cached and has an expiry on it so when the client needs to submit a new request it will re-execute the command as necessary. I basically just need to see the output to understand the 'contract' of how the data is meant to be output/consumed..then it's a pretty trivial addition to make it work.

**Comment by Daxcor69** [comment]
{"kind":"ExecCredential","apiVersion":"client.authentication.k8s.io/v1beta1
","spec":{},"status":{"expirationTimestamp":"2020-09-30T20:28:12Z","token":"5e055b19340a82fc6dd852889c1bb717ac7b0c0ea5dc806f739a1bd31ed78a89"}}



I hope this is what you need.  You know this is a dev environment, I am not
worried about exposing any secrets.  If it would be easier to do a screen
share in discord I would be cool with that.  Any thing to help the cause.



Brad



*From:* Travis Glenn Hansen [mailto:notifications@github.com]
*Sent:* Tuesday, September 29, 2020 5:11 PM
*To:* travisghansen/kubernetes-client-php
*Cc:* Daxcor69; Author
*Subject:* Re: [travisghansen/kubernetes-client-php] Noob Help please. (#3)



Yeah, I see the issue. It looks like there are 2 different variants at play
and I haven't added support for how DO does it:

   - https://banzaicloud.com/blog/kubeconfig-security/#auth-provider-plugins
   - https://banzaicloud.com/blog/kubeconfig-security/#exec-helper

Could you run the following and send over the scrubbed output?

doctl kubernetes cluster kubeconfig exec-credential --version=v1beta1
--context=default <redacted>

I'm not entirely sure what was in <redacted>.

In short what happens is a local command is executed by kubectl to get a
token/info before submitting the request to DO. Typically that data will be
cached and has an expiry on it so when the client needs to submit a new
request it will re-execute the command as necessary. I basically just need
to see the output to understand the 'contract' of how the data is meant to
be output/consumed..then it's a pretty trivial addition to make it work.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/travisghansen/kubernetes-client-php/issues/3#issuecomment-701059411>,
or unsubscribe

**Comment by travisghansen** [comment]
That might be enough..let me mess with it for a minute.

**Comment by travisghansen** [comment]
If I commit to master can you test for me?

**Comment by travisghansen** [comment]
Done: https://github.com/travisghansen/kubernetes-client-php/commit/014685ffc016e9c3b468487c975af16e99211652

Try it out and let me know if things work better.

**Comment by Daxcor69** [comment]
I will try it out tomorrow.

**Comment by Daxcor69** [comment]
I am using compose to install  your sdk so I just copy pasted your config file into my repo. It work awesome now. Thank you for taking the time to help me out. It is so nice to see developers be responsive to issues.

**Comment by travisghansen** [comment]
Great! I'll snap a new release/tag and then you can pin it sanely.
